Bugfix/fix collateral twap

gitcoindev opened the pull request:
Resolves #830

gitcoindev committed:
test: add collateral TWAP test

Author: rndquu <rndquu@gmail.com>

gitcoindev committed:
test: example of using last prices from LUSD Curve contract

Resolves: #830

molecula451 commented on packages/contracts/test/diamond/facets/UbiquityPoolFacet.t.sol:
```diff
@@ -755,4 +756,44 @@ contract UbiquityPoolFacetTest is DiamondTestSetup {
     }
+
+    function testCollateralTwap() public {
+        vm.createSelectFork("https://uk.rpc.blxrbdn.com");
```
the testing should be mocked, and not really interfacing with mainnet, this was deprecated a while ago, with #682 (where there is an open issue, to work on something that can do mainnet tests too)  If I recall, that's why all tests do not fork any mainnet, but in your testing you can use --fork-url to test stuff to mainnet on your enviroment

gitcoindev replied:
> the testing should be mocked, and not really interfacing with mainnet, this was deprecated a while ago, with #682 (where there is an open issue, to work on something that can do mainnet tests too) If I recall, that's why all tests do not fork any mainnet, but in your testing you can use --fork-url to test stuff to mainnet on your enviroment

All right, thank you for confirming this. I had the same thoughts.

molecula451 replied:
Yes, however you'll be only be able to accomplish this by running `forge` on top of the contracts folder and not on top of the `ubiquity-dollar` repo like doing it with the yarn namespace, instead you would directly using `forge` on `/contracts`

gitcoindev committed:
Merge branch 'ubiquity:development' into bugfix/fix-collateral-twap

gitcoindev committed:
feat: add LibCollateralOracle skeleton

To be used by LibUbiquityPool to fetch collateral token prices

gitcoindev committed:
feat: getLatestPrice and getDecimals for priceFeed

gitcoindev committed:
feat: emit CollateralPriceUpdated

gitcoindev committed:
Merge branch 'development' into bugfix/fix-collateral-twap

gitcoindev committed:
feat: register ChainLink price feed

gitcoindev committed:
feat: use ChainLink price feed in LibUibquityPool

gitcoindev committed:
chore: update setCollateralChainLinkPriceFeedAddress docs

gitcoindev committed:
feat: update CollateralPriceFeedSet event to emit address

gitcoindev committed:
feat: add ChainLink price feed interface functions to IUbiquityPool

gitcoindev committed:
feat: remove LibCollateralOracle

The oracle and price updates implemented in LibUbiquityPool.

gitcoindev committed:
chore: remove unused LibCollateralOracle import

gitcoindev committed:
test: implement ChainLink price feed mock

gitcoindev committed:
test: implement updateMockParams in MockChainLinkFeed

gitcoindev committed:
feat: switch to pure functcions in MockChainLinkFeed

gitcoindev committed:
feat: initialize collateralPriceFeedAddress to 0 in addCollateralToken

gitcoindev committed:
test: testCollateralPriceFeed verifies mocked ChainLink price

rndquu commented on packages/contracts/src/dollar/facets/UbiquityPoolFacet.sol:
```diff
@@ -149,6 +149,24 @@ contract UbiquityPoolFacet is IUbiquityPool, Modifiers {
+    }
+
+    /// @inheritdoc IUbiquityPool
+    function updateChainLinkCollateralPrice(
```
We could remove the `onlyAdmin` modifier because, as far as I understand, there is no issue if somebody else would fetch the latest collateral price.

Another point is that collateral prices should be updated automatically before any collateral price related operations. I couldn't find this feature in the https://github.com/ubiquity/ubiquity-dollar/blob/dac1794b832a8883fa15794eb5afa7f4054680a8/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol

gitcoindev replied:
Moved `updateChainLinkCollateralPrice` to public functions, now I need to add updates before price related operations.

gitcoindev replied:
@rndquu I found relevant `getDollarInCollateral` and `collateralUsdBalance` functions that use collateral price. I will update prices automatically in those two. Please let me know if you know other functions where the collateral price should be automatically updated.

rndquu replied:
Those are view methods for UI. I think we should update collateral prices only in `mintDollar()` and `redeemDollar()` methods.

gitcoindev replied:
Actually, `getDollarInCollateral` and `collateralUsdBalance` are view functions, the price should be updated before the call.

gitcoindev replied:
> Those are view methods for UI. I think we should update collateral prices only in `mintDollar()` and `redeemDollar()` methods.

All right! I read your comment after pasting mine after I had similar conclusions, will update `mintDollar()` and `redeemDollar()` as well.

gitcoindev replied:
done!

gitcoindev replied:
Ooops this had a side effect and messed up related tests, obviously. I need to fix those.

gitcoindev replied:
All tests fixed with mocked ChainLink price feed.

rndquu commented on packages/contracts/src/dollar/interfaces/AggregatorV3Interface.sol:
```diff
@@ -0,0 +1,37 @@
+
+pragma solidity 0.8.19;
+
+interface AggregatorV3Interface {
```
It makes sense to install https://github.com/smartcontractkit/chainlink as a dependency instead of copy-pasting the code.

rndquu commented on packages/contracts/src/dollar/interfaces/ICollateralOracle.sol:
```diff
@@ -0,0 +1,24 @@
+ * @notice Returns price quotes for all registered collaterals
+ */
+
+interface ICollateralOracle {
```
There already exists an exact interface https://github.com/ubiquity/ubiquity-dollar/blob/dac1794b832a8883fa15794eb5afa7f4054680a8/packages/contracts/src/dollar/interfaces/ITWAPOracleDollar3pool.sol

Can't we keep only 1 of them?

gitcoindev replied:
Good point, I already removed this in my local workspace.

rndquu commented on packages/contracts/src/dollar/libraries/LibUbiquityPool.sol:
```diff
@@ -598,6 +606,9 @@ library LibUbiquityPool {
         poolStorage.poolCeilings.push(poolCeiling);
+
+        // price feed
+        poolStorage.collateralPriceFeedAddresses.push(address(0));
```
Perhaps it's better to move price feed address to function args

gitcoindev committed:
chore: install ChainLink npm package

Add ChainLink contracts as project dependency.

gitcoindev committed:
feat: remove unused ICollateralOracle.sol

gitcoindev committed:
feat: switch to AggregatorV3Interface from smartcontractkit

gitcoindev committed:
feat: commit ChainLink contracts submodule

gitcoindev committed:
Merge branch 'development' into bugfix/fix-collateral-twap

molecula451 commented on packages/contracts/src/dollar/libraries/LibUbiquityPool.sol:
```diff
@@ -10,7 +10,7 @@ import {IERC20Ubiquity} from "../interfaces/IERC20Ubiquity.sol";
 import {LibAppStorage} from "./LibAppStorage.sol";
 import {LibTWAPOracle} from "./LibTWAPOracle.sol";
-import {AggregatorV3Interface} from "../../dollar/interfaces/AggregatorV3Interface.sol";
+import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
```
version naming on import path? doesn't look approachable, could you achieve it in a different way?

gitcoindev replied:
@molecula451 yes sure. I updated remapping and pushed the update.

gitcoindev committed:
feat: update ChainLink contracts remapping path

gitcoindev committed:
feat: use collateralIndex instead of collateralAddressToIndex

gitcoindev committed:
feat: move updateChainLinkCollateralPrice to public functions

Remove onlyAdmin modifier and move to public functions as discussed
during the pull request review.

gitcoindev committed:
feat: update collateral price before collateral price related functions

gitcoindev committed:
test: set up mocked ChainLink price feed in UbiquityPoolFacet test

Initialize mocked dollar price to 1_000_000.

gitcoindev committed:
test: add testCollateralPriceFeed_ShouldWorkWithMultipleCollateralTokens
test

Verify that a new collateral token and its price feed can be added.

gitcoindev committed:
test: enable second collateral in ShouldWorkWithMultipleCollateralTokens

Enable the second collateral and update its price from second feed.

gitcoindev committed:
test: add skeleton for collateral twap test with mocked meta pool

Remove forked test solution, add mocked meta pool for further
implementation.

gitcoindev committed:
test: add temporary log to testCollateralTwap

molecula451 commented on packages/contracts/src/dollar/mocks/MockChainLinkFeed.sol:
```diff
@@ -0,0 +1,74 @@
+    uint256 updatedAt;
+    uint80 answeredInRound;
+
+    constructor() {
```
this doesn't look like it's testing the implementation as diamond facets can't have constructors, all others tests, do not have constructors, can you recheck this

rndquu replied:
I'll check it

rndquu replied:
This is a mock of chainlink's data feed. It doesn't have anything in common with diamond facets because all chainlink's data feeds are deployed as separate contracts. I don't really understand what's the issue here.

gitcoindev replied:
Yes, this is just a mock to simulate the ChainLink data feed. It is used for unit tests and will not be deployed with diamond facets.

molecula451 commented on packages/contracts/src/dollar/mocks/MockChainLinkFeed.sol:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    function getRoundData(
+        uint80
```
why we should cast the value to uint80?

rndquu replied:
I'll check it

molecula451 replied:
yeah i see now this is part of the implementation

rndquu committed:
refactor: refactor UbiquityPool

rndquu committed:
Merge pull request #10 from rndquu/feat/collateral-twap

refactor: refactor UbiquityPool

molecula451 reviewed (approved):
the implementation looks good and the revamped tests are much better now, once again everyone good work here
